cmd/release-controller/http_candidate.go

```diff
@@ -0,0 +1,452 @@
+			candidate = &(releaseCandidate.ReleasePromoteJobParameters)
+		}
+		data, err := json.MarshalIndent(map[string]*ReleasePromoteJobParameters{"candidate": candidate}, "", "  ")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(data)
+		fmt.Fprintln(w)
+	}
+}
+
+// z-stream and timestamp
+var rePreviousReleaseName = regexp.MustCompile("(?P<STREAM>.*)-(?P<TIMESTAMP>[0-9]{4}-[0-9]{2}-[0-9]{2}-[0-9]{6})")
+
+func (c *Controller) findReleaseCandidates(releaseStreams ...string) (map[string]*ReleaseCandidateList, error) {
+	releaseCandidates := make(map[string]*ReleaseCandidateList)
+	if len(releaseStreams) == 0 {
+		return releaseCandidates, nil
+	}
+
+	releaseStreamTagMap, ok := c.findReleaseByName(true, releaseStreams...)
+	if !ok || len(releaseStreamTagMap) == 0 {
+		return releaseCandidates, errStreamNotFound
+	}
+
+	// next release version name for each release stream
+	next := make(map[string]*semver.Version)
+	// creation time of nightly/image that latest stable release in release stream was promoted from
+	latestPromotedTime := make(map[string]int64)
+
+	type stableRef struct {
+		from string
+		name string
+		time int64
+	}
+	stableReleases := make([]stableRef, 0)
+
+	stable, err := c.stableReleases()
+	if err != nil {
+		return releaseCandidates, err
+	}
+	for _, r := range stable.Releases {
+		for _, tag := range r.Release.Source.Spec.Tags {
+			if tag.Annotations[releaseAnnotationSource] != fmt.Sprintf("%s/%s", r.Release.Source.Namespace, r.Release.Source.Name) {
+				continue
+			}
+			if _, err := semverParseTolerant(tag.Name); err == nil {
+				t, _ := time.Parse(time.RFC3339, tag.Annotations[releaseAnnotationCreationTimestamp])
+				stableReleases = append(stableReleases,
+					stableRef{
+						from: tag.From.Name,
+						name: tag.Name,
+						time: t.Unix(),
+					})
+			}
+		}
+	}
+	sort.Slice(stableReleases, func(i, j int) bool {
+		vi, _ := semverParseTolerant(stableReleases[i].name)
+		vj, _ := semverParseTolerant(stableReleases[j].name)
+		return vi.GT(vj)
+	})
+	remaining := len(releaseStreams)
+	for _, r := range stableReleases {
+		if remaining == 0 {
+			break
+		}
+		v, _ := semverParseTolerant(r.name)
+
+		// Check if the stable version's <MAJOR>.<MINOR> matches any release stream that we are processing
+		found := false
+		for _, stream := range releaseStreams {
+			streamVersion, _ := semverParseTolerant(stream)
+			if next[stream] == nil && streamVersion.Major == v.Major && streamVersion.Minor == v.Minor {
+				found = true
+				break
+			}
+		}
+		if !found {
+			continue
+		}
+
+		// Call oc adm release info to get previous nightly info for the stable release
+		op, err := c.releaseInfo.ReleaseInfo(r.from)
+		if err != nil {
+			// releaseinfo not found, old tag
+			return releaseCandidates, fmt.Errorf("Could not get release info for tag %s: %v", r.from, err)
+		}
+
+		type releaseInfoShort struct {
```

Comment:
sorry, I meant private within the package, not the function.  generally we don't inline types except in special cases, it's confusing.